Remove sandbox extension to Web Inspector daemon in the WebContent process

pvollan opened the pull request:
#### 5f42dd57877220f5d9d574e11feb19ecf52212e9
<pre>
Remove sandbox extension to Web Inspector daemon in the WebContent process
<a href="https://rdar.apple.com/135814632">rdar://135814632</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=287347">https://bugs.webkit.org/show_bug.cgi?id=287347</a>

Reviewed by Chris Dumez.

When access to the Web Inspector daemon is blocked in the WebContent process sandbox, we cannot create
the Service Worker inspection target in the WebContent process anymore. Instead, we create the inspection
target in the UI process. We also create a bidirectional communication channel over IPC between the
Service worker in the WebContent process and the new inspection target in the UI process. The new
inspection target in the UI process will then communicate with the Web Inspector daemon and enable
inspection of the Service Worker. Messages from the Service Worker to the UI process are sent via
WebProcessProxy, and messages from the Inspector to the Service Worker are being sent through
WebSWContextManagerConnection.

* Source/JavaScriptCore/inspector/remote/RemoteInspector.h:
* Source/WebCore/workers/service/context/ServiceWorkerInspectorProxy.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
(WebKit::WebProcessProxy::createServiceWorkerDebuggable):
(WebKit::WebProcessProxy::deleteServiceWorkerDebuggable):
(WebKit::WebProcessProxy::sendMessageToInspector):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableFrontendChannel.cpp: Added.
(WebKit::ServiceWorkerDebuggableFrontendChannel::create):
(WebKit::ServiceWorkerDebuggableFrontendChannel::ServiceWorkerDebuggableFrontendChannel):
(WebKit::ServiceWorkerDebuggableFrontendChannel::sendMessageToFrontend):
* Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableFrontendChannel.h: Added..
* Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.cpp: Added.
(WebKit::ServiceWorkerDebuggableProxy::create):
(WebKit::ServiceWorkerDebuggableProxy::ServiceWorkerDebuggableProxy):
(WebKit::ServiceWorkerDebuggableProxy::connect):
(WebKit::ServiceWorkerDebuggableProxy::disconnect):
(WebKit::ServiceWorkerDebuggableProxy::dispatchMessageFromRemote):
* Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.h: Added.
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::installServiceWorker):
(WebKit::WebSWContextManagerConnection::terminateWorker):
(WebKit::WebSWContextManagerConnection::connectToInspector):
(WebKit::WebSWContextManagerConnection::disconnectFromInspector):
(WebKit::WebSWContextManagerConnection::dispatchMessageFromInspector):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.h:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.messages.in:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
* Tools/MiniBrowser/MiniBrowser.entitlements:

Canonical link: <a href="https://commits.webkit.org/290896@main">https://commits.webkit.org/290896@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/63998c2983a53d30057ab486dabf779dbfa76670

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91408 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10940 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/445 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96376 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42098 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93458 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11317 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19276 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70184 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27704 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94409 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8624 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82793 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50510 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8390 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/379 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41266 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/84212 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78710 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/385 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98379 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/90158 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18570 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79207 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18825 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78631 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78411 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22930 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/283 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11704 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14461 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18568 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23845 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/112741 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18280 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/32694 "Found 13 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.default, microbenchmarks/memcpy-wasm-medium.js.mini-mode, microbenchmarks/memcpy-wasm-small.js.dfg-eager, stress/sampling-profiler-display-name.js.dfg-eager, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-eager-jettison, wasm.yaml/wasm/stress/ipint-bbq-osr-with-try5.js.wasm-eager-jettison, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-eager, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-delegate.js.wasm-eager, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-tag-arg.js.wasm-eager, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-tag-arg.js.wasm-eager-jettison ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21739 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20046 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableFrontendChannel.h:
```diff
@@ -0,0 +1,54 @@
+    virtual ~ServiceWorkerDebuggableFrontendChannel() = default;
+
+private:
+    ServiceWorkerDebuggableFrontendChannel(WebCore::ServiceWorkerIdentifier);
```
`explicit`

cdumez commented on Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableFrontendChannel.h:
```diff
@@ -0,0 +1,54 @@
+private:
+    ServiceWorkerDebuggableFrontendChannel(WebCore::ServiceWorkerIdentifier);
+
+    ConnectionType connectionType() const override { return Inspector::FrontendChannel::ConnectionType::Remote; }
```
`final`

cdumez commented on Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableFrontendChannel.h:
```diff
@@ -0,0 +1,54 @@
+    ServiceWorkerDebuggableFrontendChannel(WebCore::ServiceWorkerIdentifier);
+
+    ConnectionType connectionType() const override { return Inspector::FrontendChannel::ConnectionType::Remote; }
+    void sendMessageToFrontend(const String& message) override;
```
`final`

cdumez commented on Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.cpp:
```diff
@@ -0,0 +1,79 @@
+    return adoptRef(*new ServiceWorkerDebuggableProxy(url, identifier, WTFMove(webProcessProxy)));
+}
+
+ServiceWorkerDebuggableProxy::ServiceWorkerDebuggableProxy(const String& url, WebCore::ServiceWorkerIdentifier identifier, WeakPtr<WebProcessProxy>&& webProcessProxy)
```
Parameter shouldn't be a WeakPtr, either pass a `WebProcessProxy&` or `WebProcessProxy*` (if nullable).

cdumez commented on Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.cpp:
```diff
@@ -0,0 +1,79 @@
+void ServiceWorkerDebuggableProxy::connect(FrontendChannel& channel, bool, bool)
+{
+    RELEASE_LOG(Inspector, "ServiceWorkerDebuggableProxy::connect");
+    if (RefPtr<WebProcessProxy> webProcessProxy = m_webProcessProxy.get())
```
Why `<WebProcessProxy>`?

cdumez commented on Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.cpp:
```diff
@@ -0,0 +1,79 @@
+void ServiceWorkerDebuggableProxy::disconnect(FrontendChannel& channel)
+{
+    RELEASE_LOG(Inspector, "ServiceWorkerDebuggableProxy::disconnect");
+    if (RefPtr<WebProcessProxy> webProcessProxy = m_webProcessProxy.get())
```
ditto.

cdumez commented on Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.cpp:
```diff
@@ -0,0 +1,79 @@
+void ServiceWorkerDebuggableProxy::dispatchMessageFromRemote(String&& message)
+{
+    RELEASE_LOG(Inspector, "ServiceWorkerDebuggableProxy::dispatchMessageFromRemote");
+    if (RefPtr<WebProcessProxy> webProcessProxy = m_webProcessProxy.get())
```
ditto.

cdumez commented on Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
```diff
@@ -592,4 +603,27 @@ void WebSWContextManagerConnection::removeNavigationFetch(WebCore::SWServerConne
+#if ENABLE(REMOTE_INSPECTOR) && PLATFORM(COCOA)
+void WebSWContextManagerConnection::connectToInspector(WebCore::ServiceWorkerIdentifier serviceWorkerIdentifier)
+{
+    RefPtr channel = ServiceWorkerDebuggableFrontendChannel::create(serviceWorkerIdentifier);
```
Seems like this returns a Ref, not a RefPtr. You can use `channel.copyRef()` on the next line if that's what you were trying to address.

cdumez commented on Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.h:
```diff
@@ -150,6 +157,9 @@ class WebSWContextManagerConnection final : public WebCore::SWContextManager::Co
     using FetchKey = std::pair<WebCore::SWServerConnectionIdentifier, WebCore::FetchIdentifier>;
     HashMap<FetchKey, Ref<WebServiceWorkerFetchTaskClient>> m_ongoingNavigationFetchTasks WTF_GUARDED_BY_CAPABILITY(m_queue.get());
+#if ENABLE(REMOTE_INSPECTOR) && PLATFORM(COCOA)
+    HashMap<WebCore::ServiceWorkerIdentifier, RefPtr<ServiceWorkerDebuggableFrontendChannel>> m_channels;
```
Can the value really be null? If not, I'd use `Ref`

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
```diff
@@ -268,5 +272,41 @@
+    RELEASE_LOG(Inspector, "WebProcessProxy::createServiceWorkerDebuggable");
+    if (!shouldEnableRemoteInspector())
+        return;
+    if (RefPtr serviceWorkerDebuggableProxy = ServiceWorkerDebuggableProxy::create(url.string(), identifier, WeakPtr { *this })) {
```
`ServiceWorkerDebuggableProxy::create()` seems to return a Ref, why store in a RefPtr and then do a null check?

pvollan replied:
Ah, good point, I have addressed this and the other issues in the latest patch.

Thanks for reviewing!

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.h:
```diff
@@ -844,7 +854,9 @@ class WebProcessProxy final : public AuxiliaryProcessProxy {
 #endif
-
+#if ENABLE(REMOTE_INSPECTOR) && PLATFORM(COCOA)
+    HashMap<WebCore::ServiceWorkerIdentifier, RefPtr<ServiceWorkerDebuggableProxy>> m_serviceWorkerDebuggableProxies;
```
Can the value really be null? If not, I'd use `Ref`.

pvollan committed:
Remove sandbox extension to Web Inspector daemon in the WebContent process
rdar://135814632
https://bugs.webkit.org/show_bug.cgi?id=287347

Reviewed by Chris Dumez.

When access to the Web Inspector daemon is blocked in the WebContent process sandbox, we cannot create
the Service Worker inspection target in the WebContent process anymore. Instead, we create the inspection
target in the UI process. We also create a bidirectional communication channel over IPC between the
Service worker in the WebContent process and the new inspection target in the UI process. The new
inspection target in the UI process will then communicate with the Web Inspector daemon and enable
inspection of the Service Worker. Messages from the Service Worker to the UI process are sent via
WebProcessProxy, and messages from the Inspector to the Service Worker are being sent through
WebSWContextManagerConnection.

* Source/JavaScriptCore/inspector/remote/RemoteInspector.h:
* Source/WebCore/workers/service/context/ServiceWorkerInspectorProxy.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
(WebKit::WebProcessProxy::createServiceWorkerDebuggable):
(WebKit::WebProcessProxy::deleteServiceWorkerDebuggable):
(WebKit::WebProcessProxy::sendMessageToInspector):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableFrontendChannel.cpp: Added.
(WebKit::ServiceWorkerDebuggableFrontendChannel::create):
(WebKit::ServiceWorkerDebuggableFrontendChannel::ServiceWorkerDebuggableFrontendChannel):
(WebKit::ServiceWorkerDebuggableFrontendChannel::sendMessageToFrontend):
* Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableFrontendChannel.h: Added..
* Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.cpp: Added.
(WebKit::ServiceWorkerDebuggableProxy::create):
(WebKit::ServiceWorkerDebuggableProxy::ServiceWorkerDebuggableProxy):
(WebKit::ServiceWorkerDebuggableProxy::connect):
(WebKit::ServiceWorkerDebuggableProxy::disconnect):
(WebKit::ServiceWorkerDebuggableProxy::dispatchMessageFromRemote):
* Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.h: Added.
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::installServiceWorker):
(WebKit::WebSWContextManagerConnection::terminateWorker):
(WebKit::WebSWContextManagerConnection::connectToInspector):
(WebKit::WebSWContextManagerConnection::disconnectFromInspector):
(WebKit::WebSWContextManagerConnection::dispatchMessageFromInspector):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.h:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.messages.in:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
* Tools/MiniBrowser/MiniBrowser.entitlements:

Canonical link: https://commits.webkit.org/290896@main
